arangod/Aql/SortingGatherExecutor.cpp

```diff
@@ -0,0 +1,323 @@
+///        ExecPhase:
+///          Fetch row of scheduled block.
+///          Pick the next (sorted) element (by strategy)
+///          Schedule this block to fetch Row
+///
+////////////////////////////////////////////////////////////////////////////////
+
+std::pair<ExecutionState, NoStats> SortingGatherExecutor::produceRow(OutputAqlItemRow& output) {
+  TRI_ASSERT(_strategy != nullptr);
+  if (!_initialized) {
+    if (_numberDependencies == 0) {
+      // We need to initialize the dependencies once, they are injected
+      // after the fetcher is created.
+      _numberDependencies = _fetcher.numberDependencies();
+      TRI_ASSERT(_numberDependencies > 0);
+      _inputRows.reserve(_numberDependencies);
+      for (size_t index = 0; index < _numberDependencies; ++index) {
+        _inputRows.emplace_back(ValueType{index});
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+        _flaggedAsDone.emplace_back(false);
+#endif
+      }
+    }
+
+    while (_dependencyToFetch < _numberDependencies) {
+      std::tie(_inputRows[_dependencyToFetch].state,
+               _inputRows[_dependencyToFetch].row) =
+          _fetcher.fetchRowForDependency(_dependencyToFetch);
+      if (_inputRows[_dependencyToFetch].state == ExecutionState::WAITING) {
+        return {ExecutionState::WAITING, NoStats{}};
+      }
+      if (!_inputRows[_dependencyToFetch].row) {
+        TRI_ASSERT(_inputRows[_dependencyToFetch].state == ExecutionState::DONE);
+        adjustNrDone(_dependencyToFetch, 4);
+      }
+      ++_dependencyToFetch;
+    }
+    _initialized = true;
+    if (_nrDone >= _numberDependencies) {
+      return {ExecutionState::DONE, NoStats{}};
+    }
+    _strategy->prepare(_inputRows);
+  } else {
+    // Activate this assert as soon as all blocks follow the done == no call api
+    // TRI_ASSERT(_nrDone < _numberDependencies);
+    if (_nrDone >= _numberDependencies) {
+      return {ExecutionState::DONE, NoStats{}};
+    }
+    if (_inputRows[_dependencyToFetch].state == ExecutionState::DONE) {
+      _inputRows[_dependencyToFetch].row = InputAqlItemRow{CreateInvalidInputRowHint()};
+      if (_nrDone >= _numberDependencies) {
+        return {ExecutionState::DONE, NoStats{}};
+      }
+    } else {
+      // This is executed on every produceRow, and will replace the row that we have returned last time
+      std::tie(_inputRows[_dependencyToFetch].state,
+               _inputRows[_dependencyToFetch].row) =
+          _fetcher.fetchRowForDependency(_dependencyToFetch);
+      if (_inputRows[_dependencyToFetch].state == ExecutionState::WAITING) {
+        return {ExecutionState::WAITING, NoStats{}};
+      }
+      if (!_inputRows[_dependencyToFetch].row) {
+        TRI_ASSERT(_inputRows[_dependencyToFetch].state == ExecutionState::DONE);
+        adjustNrDone(_dependencyToFetch, 2);
+        if (_nrDone >= _numberDependencies) {
+          return {ExecutionState::DONE, NoStats{}};
+        }
+      }
+    }
+  }
```

Comment:
There are 4 DONE checks in the code above
```cpp
        if (_nrDone >= _numberDependencies) {
          return {ExecutionState::DONE, NoStats{}};
        }
```
which I think can all safely be removed if you add it once here instead.